pkg/typesystem/resolver.go

```diff
@@ -19,73 +19,109 @@ const (
 
 // TypesystemResolverFunc is a function that implementations can implement to provide lookup and
 // resolution of a Typesystem.
-type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string, validate bool) (*TypeSystem, error)
 
-// MemoizedTypesystemResolverFunc returns a TypesystemResolverFunc that either fetches the provided authorization
-// model (if provided) or looks up the latest authorization model, and then it constructs a TypeSystem from
-// the resolved model. The type-system resolution is memoized so if another lookup of the same model occurs,
-// then the earlier TypeSystem that was constructed will be used.
-//
-// The memoized resolver function is safe for concurrent use.
-func MemoizedTypesystemResolverFunc(datastore storage.AuthorizationModelReadBackend) TypesystemResolverFunc {
-	lookupGroup := singleflight.Group{}
-
-	cache := ccache.New(ccache.Configure[*TypeSystem]())
+type TypesystemResolver interface {
+	ResolveTypesystem(ctx context.Context, storeID, modelID string, validate bool) (*TypeSystem, error)
+}
 
-	return func(ctx context.Context, storeID, modelID string) (*TypeSystem, error) {
-		ctx, span := tracer.Start(ctx, "MemoizedTypesystemResolverFunc")
-		defer span.End()
+type cachedTypesystemResolver struct {
+	modelReader  storage.AuthorizationModelReadBackend
+	cache        *ccache.Cache[*TypeSystem]
+	singleflight singleflight.Group
+}
 
-		var err error
+type TypesystemResolverOption func(c *cachedTypesystemResolver)
 
-		if modelID != "" {
-			if _, err := ulid.Parse(modelID); err != nil {
-				return nil, ErrModelNotFound
-			}
-		}
+func WithTypesystemResolverCache(cache *ccache.Cache[*TypeSystem]) TypesystemResolverOption {
+	return func(c *cachedTypesystemResolver) {
+		c.cache = cache
+	}
+}
 
-		if modelID == "" {
-			v, err, _ := lookupGroup.Do(fmt.Sprintf("FindLatestAuthorizationModelID:%s", storeID), func() (interface{}, error) {
-				return datastore.FindLatestAuthorizationModelID(ctx, storeID)
-			})
-			if err != nil {
-				if errors.Is(err, storage.ErrNotFound) {
-					return nil, ErrModelNotFound
-				}
+func NewCachedTypesystemResolver(
+	modelReader storage.AuthorizationModelReadBackend,
+	opts ...TypesystemResolverOption,
+) *cachedTypesystemResolver {
+	resolver := &cachedTypesystemResolver{
+		modelReader: modelReader,
+	}
 
-				return nil, fmt.Errorf("failed to FindLatestAuthorizationModelID: %w", err)
-			}
+	for _, opt := range opts {
+		opt(resolver)
+	}
 
-			modelID = v.(string)
-		}
+	if resolver.cache == nil {
+		resolver.cache = ccache.New(ccache.Configure[*TypeSystem]())
+	}
 
-		key := fmt.Sprintf("%s/%s", storeID, modelID)
+	return resolver
+}
 
-		item := cache.Get(key)
-		if item != nil {
-			return item.Value(), nil
+// ResolveTypesystem resolves
```

Comment:
We should document here all this function does:

- If model Id is given, try to fetch that model
- If model is not given, try to fetch the latest model for that store
- it validates the model